### ISSUE TITLE
Images module makes it difficult to differentiate for CSS2

### ISSUE THREAD

**Issue by alanhogan** [Compass/compass#309]
Consider:

```
background: #f5e0e0 none;
// Override bg for CSS3 browser
@include background( transparent linear-gradient(
    rgba(98%, 74%, 74%, 0.0), 
    rgba(98%, 74%, 74%, 1.0)
));
```

The output is: 

```
background: #f5e0e0 none; 
background: transparent ; 
background: transparent url('data:image/svg+xml;base64,PD9…z4g'); 
background: transparent -webkit-gradient(linear, 50% 0%, 50% 100%, color-stop(0%, rgba(249, 188, 188, 0)), color-stop(100%, #f9bcbc)); 
background: transparent -moz-linear-gradient(rgba(249, 188, 188, 0), #f9bcbc);
background: transparent linear-gradient(rgba(249, 188, 188, 0), #f9bcbc); }
```

So CSS2 browsers see `background: transparent` now! Obviously this can cause a problem… and the “fix” would be to pass  `background()` the color `#f5e0e0` but this actually changes the meaning of the CSS3 styles quite a bit.

**Comment by chriseppstein** [comment]
It is a conundrum. I have an idea.

**Comment by alanhogan** [comment]
Great, looking forward to it… my naïve hope would be for the images module to simply stop emitting the plain “`background: <color>;`” line altogether.

**Comment by chriseppstein** [comment]
Don&#39;t output transparent in css2 ouput.

Closed by 119ce9d018ef25bf08531114ce22307306dfc212

**Comment by chriseppstein** [comment]
Now that same input will print out:

```
background: #f5e0e0 none; 
background: ; 
background: transparent url('data:image/svg+xml;base64,PD9…z4g'); 
background: transparent -webkit-gradient(linear, 50% 0%, 50% 100%, color-stop(0%, rgba(249, 188, 188, 0)), color-stop(100%, #f9bcbc)); 
background: transparent -moz-linear-gradient(rgba(249, 188, 188, 0), #f9bcbc);
background: transparent linear-gradient(rgba(249, 188, 188, 0), #f9bcbc); }
```

Messy but functional.

**Comment by alanhogan** [comment]
Cheers. Seems like this would still make it quite difficult, or impossible, to specify different non-transparent background colors for CSS2 and CSS3, respectively, though. Thoughts?

**Comment by chriseppstein** [comment]
I don't understand why someone would want to do that.

**Comment by chriseppstein** [comment]
I reverted this changeset, `transparent` is a [legal css2 value](http://www.w3.org/TR/CSS2/colors.html) -- I thought it was a css3 thing. I don't know why you want to use different colors in these circumstances, but I think the right thing is to specify the same color you want in all cases or don't use the shorthand notation or don't use the compass mixin.

**Comment by alanhogan** [comment]
Here is an example. Consider invalid form items. With CSS3 browsers, I
could use a semitransparent background gradient (maybe the page has a
patterned/noise background) or box-shadow to differentiate it, but
need to do something else for CSS2 browsers. A light red background
would be the way to do that. So now you have a case where the
background color needs to be different for different browsers and this
mixin becomes unusable. Which is sad, because it's fantastically
_useful._

On Monday, March 28, 2011, chriseppstein
reply@reply.github.com
wrote:

> I reverted this changeset, `transparent` is a [legal css2 value](http://www.w3.org/TR/CSS2/colors.html) -- I thought it was a css3 thing. I don't know why you want to use different colors in these circumstances, but I don't think the right thing is to specify the same color you want in all cases or don't use the shorthand notation or don't use the compass mixin.
> 
> ## 
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/chriseppstein/compass/issues/309#comment_926855

## 

Alan Hogan

**Comment by chriseppstein** [comment]
@alanhogan, I think you should install modernizr and scope your rules according to feature detection or do one of the other remedies I discussed above. This mixin has a very clear and concise behavior and I think your needs are not common. If you look at the implementation of this mixin you'll see it's not doing a whole lot, please take its guts and modify them for your needs on your projects.

**Comment by mirisuzanne** [comment]
It's actually a very common use case, and I had also assumed that the background mixin would work with it. Sometimes the color I use with a gradient will not work as a legible color without the gradient.

Modernizr is certainly an option, but @alanhogan's approach is the accepted CSS best-practice and I'm actually not clear on why the background mixin would output that first line in any case. That line is so different from what was input that I don't know why I would ever want it. Can you explain why it's there?

**Comment by chriseppstein** [comment]
The source of the first line is compass's attempt to create a fallback value for css2 browsers. Maybe compass is doing too much here, but I thought we'd go a little deeper down this rabbit hole before we turn back, so I made a new compass function called `css2-fallback` that you can use to pass a value and a css2 fallback as a single value. If people hate this then we can stop trying to be smart. So now if you pass this:

```
@include background(css2-fallback(transparent, #f5e0e0) linear-gradient(
    rgba(98%, 74%, 74%, 0.0), 
    rgba(98%, 74%, 74%, 1.0)
));
```

You'll get this:

```
background: #f5e0e0; 
background: transparent url('data:image/svg+xml;base64,PD9…z4g'); 
background: transparent -webkit-gradient(linear, 50% 0%, 50% 100%, color-stop(0%, rgba(249, 188, 188, 0)), color-stop(100%, #f9bcbc)); 
background: transparent -moz-linear-gradient(rgba(249, 188, 188, 0), #f9bcbc);
background: transparent linear-gradient(rgba(249, 188, 188, 0), #f9bcbc); }
```

**Comment by mirisuzanne** [comment]
I think that yes, Compass is doing too much in the initial approach. Because what it is doing is not a default behavior and not a reliable fallback behavior. 

Now this, on the other hand, is an interesting proposal. I would adjust it a bit, because I'm not sure why both values are being passed to the function. I'd keep everything explicit and run it like this:

```
@include background(css2-fallback(#f5e0e0) transparent linear-gradient(
    rgba(98%, 74%, 74%, 0.0), 
    rgba(98%, 74%, 74%, 1.0)
));
```

and then that css2 fallback line would only appear in the css if it is passed the css2-fallback() function, and I would pass it exactly what the function has been given. Less magic, even more function.

**Comment by chriseppstein** [comment]
The reason that it takes two values is that transparent would be in the output otherwise, conceptually they are a single value with different representations in different contexts.

**Comment by chriseppstein** [comment]
Closed by f2ceec4680a2305bf716c5eba2fbc52768ff721e. Try this and let me know if you hate it and would rather manage legacy browsers outside of the css3 mixins.

**Comment by mirisuzanne** [comment]
That makes some technical sense in the background, but it just doesn't match how these tools are used or the expected behavior of either Compass or CSS. In reality I want completely separate control of my fallback, and I will probably even write it at a different time (durring browser-testing). It makes more sense to me that I would write my css2 shorthand and my css3 shorthand each once, separately, rather than duplicating each sub-property along with its fallback in one long mixed css2 and css3 combined list.

This is all too magic for my taste, with Compass trying to determine what I want as a fallback and I have to override if I want something else. It just isn't reliable, and doesn't follow the way we are used to writing CSS. If we want to add the ability for my css2 fallback to be passed in the same mixin as my css3, but as its own list, I can get behind that but don't care much. Otherwise, I'd take it out entirely.

**Comment by chriseppstein** [comment]
I agree with you on a certain level. I can see how this would get in the way for users like yourself, but I think it's a pretty good behavior for n00bies who don't really grok the whole progressive enhancement concept yet.

What if we added a config variable to turn off css2 fallbacks in the css3 module?

```
$css2-degradation: true !default;
```

**Comment by mirisuzanne** [comment]
you might be right, and that solution would be acceptable for what I need, but my experience with n00bies is that they are actually even more confused with magic solutions than I am - especially when the abstraction is so leaky. They will expect a solution that is fool-proof or none at all. I don't think we can do foolproof, so I vote for none.

**Comment by chriseppstein** [comment]
It's annoying that you're right.

**Comment by alanhogan** [comment]
Chris, I sympathize with your intention and the disappointment that comes with a good idea being perhaps less than ideal in practice. All I can say is, thanks for caring so much and listening so well.

**Comment by alanhogan** [comment]
Also, I thought I was replying by email, but I don’t see my comments here. Basically, I think @ericam expressed a lot of my own views very well (thanks).

**Comment by chriseppstein** [comment]
I'm not quite sure how to handle the v0.11 early adopters of the new css3 mixins. If they are already relying on the fallback behavior. Any suggestions?

**Comment by mirisuzanne** [comment]
I'd guess at this point anyone adopting the Beta will see a note in the next changelog and make the appropriate adjustments. That's the beauty of Beta. Or maybe I'm just a hard-hearted developer.

**Comment by chriseppstein** [comment]
I moved the simple bg support to a new mixin called background-with-css2-fallback which has the same behavior as the current bg mixin. The release notes mention this, but those who didn't adopt the prerelease will be none the wiser. I'll push this change in a little bit.

**Comment by alanhogan** [comment]
That sounds like the best way forward to me.

**Comment by mirisuzanne** [comment]
closed by: 0eef08e6ce664d5d905928c7a7c4e66f80e5e8c7

**Comment by alanhogan** [comment]
I’m really glad to see this fixed… I recently noticed my IE visitors were seeing white text on white background in my site header, with this issue as the cause (despite my creation of a fallback color in the source SCSS) — updating to the latest beta (or specifying a color for the images model) fixed my issue.  You did the right thing here, Chris.
